### DISCUSSION TITLE
Impossible to connect a simple relays to ESP32 DevKit v.1 (WROOM-ESP32)

### DISCUSSION THREAD

**Discussion by nastiliano** [micropython/micropython#11789] [Hardware & Peripherals]
Hello everybody.

I'm simply atonished, because I've been try to connect simply a relays to my ESP32 and they are not working properly.

The  GPIO pins are numbers: 14 and 27. I've tried to configure them in all possible modes...

Here are my schemes: (the universal way to connect a relay, nothing more sofisticated... I've tried NPN and PNP options) :

![WhatsApp Image 2023-06-15 at 17 06 59](https://github.com/micropython/micropython/assets/115594508/d4bd3b24-9481-4960-805a-3640a91bfe32)
![WhatsApp Image 2023-06-15 at 17 06 34](https://github.com/micropython/micropython/assets/115594508/0ae0e60b-80c5-498f-b94e-a6159822df50)

Any ideas please??

Thank you very much.

**Comment by robert-hh** [comment]
A few assumptions:
a) You have set the GPIO to OUT mode
b) You use a 5V relais.

Q: How much current does the relais take to switch?
Q: What are the values of the resistors
Q: Which transistor do you use.
Note about the circuits:
1. With Optocoupler.
Optocouplers have a poor current transfer rate. For the 4N35 it's about 1. That means, if you drive the LED with 1mA, you can switch 1 mA at the output side to saturation. If the relais need 100mA, you  would have to drive the LED with 100mA, which both the LED and the ESP32 do not like. So you would need another transistor driven by the optocoupler.

2. With PNP-Transistor.
There the current transfer rate is in the range if about 100 (rule of thumb), so the relais may switch on, but not off. That's because the output voltage of the ESP32 is 3.3V, the emitter is at 5V, and so there is always current through the base, keeping the transistor somewhat switched. The circuit should work when you use a NPN transistor to GND, which can handle the current required for the relais.

**Reply by 2dof** [reply]
Read about [transistor switching relay configuration](https://www.electronics-tutorials.ws/blog/relay-switch-circuit.html)

also [Basic ESP Circuit - Mechanical Relays](https://www.letscontrolit.com/wiki/index.php/Basics:_Relays)

**Reply by nastiliano** [reply]
@robert-hh  I'm going to answer you:

a) Of course
b) Of course, even I tried 3VDC relays

Q1: I think this is the problem. I poor coil current.
Q2: They were 330ohm, but even I tried without resistors in all options 0ohm
Q3: I tried PC817 opto, npn BC33725 and pnp BC640

1 and 2: Mmmmm.... I see. And what is a solution for ESP32?? A darlington or specific configuration??? I tried a lot and is frustrating

thank you very much

**Reply by nastiliano** [reply]
@2dof 
Ok, thanks, this night I'm going to read it

**Reply by 2dof** [reply]
also check your relay circuit without  controlling via GPIO  -> just  connect to GND/Vcc according to solution.

**Comment by nastiliano** [comment]
SOLVED:

It was hard... But finally works!

The only configuration that it was good for me was "NPN Darlington Configuration", no otrer configuration: no PNP Darlington, no NPN Emitter Darlington configuration...

Thanks to everybody.
